xds/src/main/java/io/grpc/xds/client/XdsClientImpl.java

```diff
@@ -747,4 +986,116 @@ private void notifyWatcher(ResourceWatcher<T> watcher, T update) {
     }
   }
 
+  private class ResponseHandler implements XdsResponseHandler {
+    final ServerInfo serverInfo;
+
+    ResponseHandler(ServerInfo serverInfo) {
+      this.serverInfo = serverInfo;
+    }
+
+    @Override
+    public void handleResourceResponse(
+        XdsResourceType<?> xdsResourceType, ServerInfo serverInfo, String versionInfo,
+        List<Any> resources, String nonce, ProcessingTracker processingTracker) {
+      checkNotNull(xdsResourceType, "xdsResourceType");
+      syncContext.throwIfNotInThisSynchronizationContext();
+      Set<String> toParseResourceNames =
+          xdsResourceType.shouldRetrieveResourceKeysForArgs()
+          ? getResourceKeys(xdsResourceType)
+          : null;
+      XdsResourceType.Args args = new XdsResourceType.Args(serverInfo, versionInfo, nonce,
+          bootstrapInfo, securityConfig, toParseResourceNames);
+      handleResourceUpdate(args, resources, xdsResourceType, processingTracker);
+    }
+
+    @Override
+    public void handleStreamClosed(Status status, boolean inRetry,
+                                   ControlPlaneClient cpcThatClosed) {
+      syncContext.throwIfNotInThisSynchronizationContext();
+      if (!inRetry) {
+        cleanUpResourceTimers(cpcThatClosed);
+      }
+      if (cpcThatClosed.isResponseRecieved()) {
+        return; // Not considered an error
+      }
+
+      Status error = status.isOk() ? Status.UNAVAILABLE.withDescription(CLOSED_BY_SERVER) : status;
+      boolean checkForFallback = !status.isOk() && !inRetry;
+
+      List<String> authoritiesForClosedCpc = getActiveAuthorities(cpcThatClosed);
+      for (Map<String, ResourceSubscriber<? extends ResourceUpdate>> subscriberMap :
+          resourceSubscribers.values()) {
+        for (ResourceSubscriber<? extends ResourceUpdate> subscriber : subscriberMap.values()) {
+          if (subscriber.hasResult() || !authoritiesForClosedCpc.contains(subscriber.authority)) {
+            continue;
+          }
+
+          // try to fallback to lower priority control plane client
+          if (checkForFallback
+              && doFallbackForAuthority(
+                  cpcThatClosed, serverInfo, subscriber.serverInfos, subscriber.authority)) {
```

Comment:
The docs for `doFallbackForAuthority()` say server info should be for the active CPC. But here it is identical to `cpcThatClosed.getServerInfo()`.

Reply:
Since we are only trying to do fallback for authorities for which the closed cpc was the active cpc,  they are the same.  However, since they are the same, and this is the only place that this method is called, there is no reason to pass the serverInfo.

```
authoritiesForClosedCpc = getActiveAuthorities(cpcThatClosed)
...
          if (subscriber.hasResult() || !authoritiesForClosedCpc.contains(subscriber.authority)) {
            continue;
          }
```

Reply:
Logic was moved to manageControlPlaneClient during our pair programming session.